core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultTopologicalOrderGraph.java

```diff
@@ -0,0 +1,152 @@
+        AtomicInteger index = new AtomicInteger(1);
+        AtomicInteger stackIndex = new AtomicInteger(0);
+        int size = forwardEdges.length;
+        int[] stack = new int[size];
+        int[] indexMap = new int[size];
+        int[] lowMap = new int[size];
+        boolean[] onStackSet = new boolean[size];
+        List<BitSet> components = new ArrayList<>();
+        componentMap.clear();
+
+        for (int node = 0; node < size; node++) {
+            if (indexMap[node] == 0) {
+                strongConnect(node, index, stackIndex, stack, indexMap, lowMap, onStackSet, components);
+            }
+        }
+
+        int ordIndex = 0;
+        for (int i = components.size() - 1; i >= 0; i--) {
+            var component = components.get(i);
+            var componentNodes = new ArrayList<Integer>(component.cardinality());
+            for (int node = component.nextSetBit(0); node >= 0; node = component.nextSetBit(node + 1)) {
+                ord[node] = ordIndex;
+                componentNodes.add(node);
+                componentMap.put(node, componentNodes);
+                ordIndex++;
+                if (node == Integer.MAX_VALUE) {
+                    break;
+                }
+            }
+        }
+    }
+
+    private void strongConnect(int node, AtomicInteger index, AtomicInteger stackIndex, int[] stack,
+            int[] indexMap,
+            int[] lowMap, boolean[] onStackSet, List<BitSet> components) {
+        // Set the depth index for node to the smallest unused index
+        indexMap[node] = index.get();
+        lowMap[node] = index.get();
+        index.getAndIncrement();
+        stack[stackIndex.getAndIncrement()] = node;
+        onStackSet[node] = true;
+
+        // Consider successors of node
+        for (int successor : forwardEdges[node]) {
+            if (indexMap[successor] == 0) {
+                // Successor has not yet been visited; recurse on it
+                strongConnect(successor, index, stackIndex, stack, indexMap, lowMap, onStackSet, components);
+                lowMap[node] = Math.min(lowMap[node], lowMap[successor]);
+            } else if (onStackSet[successor]) {
+                // Successor is in stack S and hence in the current SCC
+                // If successor is not on stack, then (node, successor) is an edge pointing to an SCC already found and must be ignored
+                // The next line may look odd - but is correct.
+                // It says successor.index not successor.low; that is deliberate and from the original paper
+                lowMap[node] = Math.min(lowMap[node], indexMap[successor]);
+            }
+        }
+
+        // If node is a root node, pop the stack and generate an SCC
+        if (onStackSet[node] && lowMap[node] == indexMap[node]) {
+            BitSet out = new BitSet();
+
+            int current;
+            do {
+                current = stack[stackIndex.decrementAndGet()];
+                onStackSet[current] = false;
```

Comment:
I don't understand - if `out` is a `BitSet`, why is `onStackSet` (which also stores bits) not a `BitSet` as well?

Reply:
onStackSet is simply a mapping from node (an int)  and does not need any of the BitSet methods. Essentially, indexMap, lowMap, and onStackSet acts as additional attributes on nodes.

out is a BitSet since we need to iterate it and access other BitSet methods.